File: contracts/api/clients/jakarta/src/main/java/app/epistola/client/jakarta/identity/ClientIdentity.java
// SPDX-FileCopyrightText: Epistola Nederland B.V.
//
// SPDX-License-Identifier: EUPL-1.2

package app.epistola.client.jakarta.identity;

import jakarta.ws.rs.client.ClientRequestContext;
import jakarta.ws.rs.client.ClientRequestFilter;
import jakarta.ws.rs.core.HttpHeaders;
import java.io.IOException;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

/**
 * Manages the client identity headers ({@code User-Agent} and {@code X-EP-Node-Id}) the contract
 * requires on every Epistola API request.
 *
 * <p>The {@code User-Agent} always starts with {@code epistola-contract/{contractVersion}}.
 * Additional product tokens can be added to describe the full software stack.
 *
 * <pre>{@code
 * ClientIdentity identity = ClientIdentity.builder()
 *     .nodeId("my-pod-123")
 *     .product("zaakafhandelcomponent", "3.4.0")
 *     .build();
 *
 * GenerationApi api = RestClientBuilder.newBuilder()
 *     .baseUri(URI.create("https://epistola.example.com/api"))
 *     .register(identity.filter())
 *     .build(GenerationApi.class);
 * }</pre>
 *
 * Produces headers:
 * <pre>
 * User-Agent: epistola-contract/1.1.0 zaakafhandelcomponent/3.4.0
 * X-EP-Node-Id: my-pod-123
 * </pre>
 */
public final class ClientIdentity {

    /**
     * Header carrying the node identifier the contract requires, from the spec's
     * {@code x-client-identity} registry.
     */
    public static final String HEADER_NODE_ID = ContractIdentity.NODE_ID_HEADER;

    static final String CONTRACT_PRODUCT = ContractIdentity.CONTRACT_PRODUCT;

    private static final String CONTRACT_VERSION_RESOURCE = "/epistola-contract-version.txt";

    private static final class VersionHolder {
        private static final String VALUE = readContractVersion();

        private static String readContractVersion() {
            try (var stream = ClientIdentity.class.getResourceAsStream(CONTRACT_VERSION_RESOURCE)) {
                if (stream == null) {
                    return "unknown";
                }
                return new String(stream.readAllBytes(), StandardCharsets.UTF_8).trim();
            } catch (IOException e) {
                return "unknown";
            }
        }
    }

    private final String userAgent;
    private final String nodeId;

    private ClientIdentity(String userAgent, String nodeId) {
        this.userAgent = userAgent;
        this.nodeId = nodeId;
    }

    /**
     * The contract version this client library was built against, read from the bundled
     * {@code epistola-contract-version.txt} resource the build writes from the spec.
     */
    public static String contractVersion() {
        return VersionHolder.VALUE;
    }

    public static Builder builder() {
        return new Builder();
    }

    public String getUserAgent() {
        return userAgent;
    }

    public String getNodeId() {
        return nodeId;
    }

    /**
     * A JAX-RS request filter that sets {@code User-Agent} and {@code X-EP-Node-Id} on every
     * outgoing request. Register it on a {@code RestClientBuilder}, or let
     * {@code EpistolaRestClientListener} register it for you.
     */
    public ClientRequestFilter filter() {
        return new ClientIdentityFilter(this);
    }

    /** Builds a {@link ClientIdentity}; every product token is validated as it is added. */
    public static final class Builder {

        private final List<String> products = new ArrayList<>();
        private String nodeId;

        /**
         * Sets the node identifier (Kubernetes pod name, container ID, hostname). Defaults to the
         * local hostname.
         */
        public Builder nodeId(String nodeId) {
            // A blank node id would send the mandatory header empty, which is worse than the
            // hostname default it would be replacing.
            if (nodeId != null && nodeId.isBlank()) {
                throw new IllegalArgumentException("nodeId must not be blank");
            }
            this.nodeId = nodeId;
            return this;
        }

        /**
         * Adds a product/version pair to the {@code User-Agent}. Products appear in the order they
         * are added, after the {@code epistola-contract/{version}} token.
         */
        public Builder product(String name, String version) {
            if (name == null || name.isBlank()) {
                throw new IllegalArgumentException("Product name must not be blank");
            }
            if (version == null || version.isBlank()) {
                throw new IllegalArgumentException("Product version must not be blank");
            }
            if (name.indexOf('/') >= 0 || name.indexOf(' ') >= 0) {
                throw new IllegalArgumentException("Product name must not contain '/' or spaces");
            }
            products.add(name + ContractIdentity.VERSION_SEPARATOR + version);
            return this;
        }

        public ClientIdentity build() {
            StringBuilder userAgent = new StringBuilder(CONTRACT_PRODUCT)
                    .append(ContractIdentity.VERSION_SEPARATOR)
                    .append(contractVersion());
            for (String product : products) {
                userAgent.append(ContractIdentity.PRODUCT_SEPARATOR).append(product);
            }
            return new ClientIdentity(userAgent.toString(), nodeId != null ? nodeId : localHostname());
        }

        private static String localHostname() {
            try {
                return InetAddress.getLocalHost().getHostName();
            } catch (UnknownHostException e) {
                // A container with no resolvable hostname is normal; the header is still required.
                return "unknown-host";
            }
        }
    }

    private static final class ClientIdentityFilter implements ClientRequestFilter {

        private final ClientIdentity identity;

        private ClientIdentityFilter(ClientIdentity identity) {
            this.identity = identity;
        }

        @Override
        public void filter(ClientRequestContext requestContext) {
            requestContext.getHeaders().putSingle(HttpHeaders.USER_AGENT, identity.userAgent);
            requestContext.getHeaders().putSingle(HEADER_NODE_ID, identity.nodeId);
        }
    }
}

File: client-kotlin-spring-restclient/client/src/main/kotlin/app/epistola/client/identity/ClientIdentity.kt
package app.epistola.client.identity

import org.springframework.http.HttpRequest
import org.springframework.http.client.ClientHttpRequestExecution
import org.springframework.http.client.ClientHttpRequestInterceptor
import org.springframework.http.client.ClientHttpResponse
import java.net.InetAddress

/**
 * Manages client identity headers (`User-Agent` and `X-EP-Node-Id`) for all Epistola API requests.
 *
 * The `User-Agent` always starts with `epistola-contract/{contractVersion}`.
 * Additional product tokens can be added to describe the full software stack.
 *
 * Example:
 * ```
 * val identity = ClientIdentity.builder()
 *     .nodeId("my-pod-123")
 *     .product("valtimo-epistola-plugin", "1.2.0")
 *     .product("gzac", "5.0.0")
 *     .build()
 *
 * val restClient = RestClient.builder()
 *     .baseUrl("http://localhost:8080/api")
 *     .requestInterceptor(identity.interceptor())
 *     .build()
 * ```
 *
 * Produces headers:
 * ```
 * User-Agent: epistola-contract/0.3.0 valtimo-epistola-plugin/1.2.0 gzac/5.0.0
 * X-EP-Node-Id: my-pod-123
 * ```
 */
class ClientIdentity private constructor(
    val userAgent: String,
    val nodeId: String,
) {
    companion object {
        const val HEADER_NODE_ID = "X-EP-Node-Id"
        internal const val CONTRACT_PRODUCT = "epistola-contract"

        /**
         * The contract version this client library was built against.
         * Read from the bundled `epistola-contract-version.txt` resource at build time.
         */
        val contractVersion: String by lazy {
            ClientIdentity::class.java.getResourceAsStream("/epistola-contract-version.txt")
                ?.bufferedReader()?.readLine()?.trim()
                ?: "unknown"
        }

        fun builder(): Builder = Builder()
    }

    /**
     * Creates a Spring [ClientHttpRequestInterceptor] that adds the `User-Agent`
     * and `X-EP-Node-Id` headers to every outgoing request.
     */
    fun interceptor(): ClientHttpRequestInterceptor = ClientIdentityInterceptor(this)

    class Builder {
        private var nodeId: String? = null
        private val products = mutableListOf<Pair<String, String>>()

        /**
         * Sets the node identifier (e.g. Kubernetes pod name, hostname).
         * If not set, defaults to the local hostname.
         */
        fun nodeId(nodeId: String) = apply { this.nodeId = nodeId }

        /**
         * Adds a product/version pair to the `User-Agent` header.
         * Products are appended in the order they are added, after the
         * `epistola-contract/{version}` token.
         */
        fun product(name: String, version: String) = apply {
            require(name.isNotBlank()) { "Product name must not be blank" }
            require(version.isNotBlank()) { "Product version must not be blank" }
            require(!name.contains('/') && !name.contains(' ')) {
                "Product name must not contain '/' or spaces"
            }
            products.add(name to version)
        }

        fun build(): ClientIdentity {
            val tokens = mutableListOf("$CONTRACT_PRODUCT/$contractVersion")
            products.forEach { (name, version) -> tokens.add("$name/$version") }

            return ClientIdentity(
                userAgent = tokens.joinToString(" "),
                nodeId = nodeId ?: InetAddress.getLocalHost().hostName,
            )
        }
    }

    private class ClientIdentityInterceptor(
        private val identity: ClientIdentity,
    ) : ClientHttpRequestInterceptor {
        override fun intercept(
            request: HttpRequest,
            body: ByteArray,
            execution: ClientHttpRequestExecution,
        ): ClientHttpResponse {
            request.headers.set("User-Agent", identity.userAgent)
            request.headers.set(HEADER_NODE_ID, identity.nodeId)
            return execution.execute(request, body)
        }
    }
}
